mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java

```diff
@@ -60,25 +69,51 @@
         }
     }
 
-    private List<Attachment> listAttachments(Multipart multipart) throws IOException {
-        ImmutableList.Builder<Attachment> attachments = ImmutableList.builder();
+    private List<MessageAttachment> listAttachments(Multipart multipart) throws IOException {
+        ImmutableList.Builder<MessageAttachment> attachments = ImmutableList.builder();
         MessageWriter messageWriter = new DefaultMessageWriter();
         for (Entity entity : multipart.getBodyParts()) {
-            if (isAttachment(entity)) {
-                Optional<ContentTypeField> contentTypeField = contentTypeField(entity.getHeader().getField(CONTENT_TYPE));
-                Optional<String> contentType = contentType(contentTypeField);
-                Optional<String> name = name(contentTypeField);
-                
-                attachments.add(Attachment.builder()
-                        .bytes(getBytes(messageWriter, entity.getBody()))
-                        .type(contentType.or(DEFAULT_CONTENT_TYPE))
-                        .name(name)
-                        .build());
+            if (entity.isMultipart()) {
+                attachments.addAll(listAttachments((Multipart) entity.getBody()));
```

Comment:
Unfortunately entity.isMultipart doesn't check that the type of the body is Multipart. Can you make an explicit check before casting?